Fix missing example project link on Core Dep Management Doc in User Manual (#24954)
gradle/gradle: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/core_dependency_management.adoc

[[sec:dependency-mgmt-in-gradle]]
== Dependency Management in Gradle

.Dependencies management at a glance
image::dependency-management-resolution.png[]

Gradle has built-in support for dependency management.

Let's explore the main concepts with the help of a theoretical but common project:

* This project builds Java source code.
* Some Java source files import classes from the link:https://github.com/google/guava[Google Guava] library.
* This project uses link:http://junit.org/junit5/[JUnit] for testing.

> The `build.gradle` file might look as follows:

Author: Fixed in new commit.

.Gradle build file with dependencies
====
include::sample[dir="snippets/dependencyManagement/introductionCoreDependencies/kotlin",files="build.gradle.kts"]
include::sample[dir="snippets/dependencyManagement/introductionCoreDependencies/groovy",files="build.gradle"]
====

> <1> *Define repositories for project.*

Reviewer: I find myself wanting to add some sort of implied prefix to these bullet points, but it doesn't work consistently for all points.  For instance I might expect:

_Here we_ Define repositories for the project.
OR
_This is how you would_ Define repositories for the project.

But this doesn't work well with the second bullet point "Remote and local repositories for dependency locations" which I would prefix with something like _this is an example of_ or _this is how you would configure_.

I wonder if we could make these read more consistently in a single voice?  e.g.

_Here we..._
Define all repositories for the project.
Declare each remote and local repository for dependency locations.
OR
_This is how you would configure..._
The container of all repositories of the project
An individual remote or local repository for dependency locations

Author: Fixed in final commit.

> <1> *Define artifact sources for project.*

Author: Fixed in latest commit.

<2> *Remote and local repositories for artifact locations.*
+
You can <<declaring_repositories.adoc#declaring-repositories,declare _repositories_>> to tell Gradle where to fetch local or remote _dependencies_. +

> In this example, Gradle fetches _dependencies_ from the link:https://repo1.maven.org/maven2/[Maven] and link:https://maven.google.com/[Google] _repositories_. +

Author: Fixed in latest commit.

During a build, Gradle locates and downloads the dependencies, a process called <<dependency_resolution.adoc#sec:how-gradle-downloads-deps,_dependency resolution_>>.
Gradle then <<dependency_resolution.adoc#sec:dependency_cache,stores resolved dependencies in a local cache>> called the _dependency cache_.

> Subsequent builds use the cache to avoid unnecessary network calls and speed up the build process.

Reviewer: ```suggestion
Subsequent builds use this cache to avoid unnecessary network calls and speed up the build process.
```

Author: Fixed in final commit.

<3> *Define dependencies used by project.*

> <4> *Specific dependency name and version within a specific scope.*

Reviewer: The repetition of "specific" feels awkward to me.

Perhaps something like:
```suggestion
<4> *Declare unique dependency names and versions for specific scopes.*
```

Author: Fixed in final commit.

[[sec:dependency-mgmt-in-gradle]]
== Dependency Management in Gradle

.Dependencies management at a glance
image::dependency-management-resolution.png[]

Gradle has built-in support for dependency management.

Let's explore the main concepts with the help of a theoretical but common project:

* This project builds Java source code.
* Some Java source files import classes from the link:https://github.com/google/guava[Google Guava] library.
* This project uses link:http://junit.org/junit5/[JUnit] for testing.

The `build.gradle` file might look as follows:

.Gradle build file with dependencies
====
include::sample[dir="snippets/dependencyManagement/introductionCoreDependencies/kotlin",files="build.gradle.kts"]
include::sample[dir="snippets/dependencyManagement/introductionCoreDependencies/groovy",files="build.gradle"]
====

> <1> Define artifact sources for project

Reviewer 2: The terminology used in the call outs does not match the rest of the doc.

Given that you effectively explain the different pieces below, I would suggest to:
* Either move the sentences below to the right callout
* Or remove the callouts altogether and rely on the paragraphs below to explain the snippet.

Author: > Either move the sentences below to the right callout
Chose this in the latest commit.

> <2> Remote and local repositories for artifact locations (link:https://mvnrepository.com/repos/central[Maven] and link:https://maven.google.com/[Google] repositories)

Reviewer 2: The links are inconsistent. We should use the same kind of links for both:
* Either replace the Google one with https://mvnrepository.com/repos/google
* Or use https://repo1.maven.org/maven2/ for Central

I am in favor of the second option.

Author: > Or use https://repo1.maven.org/maven2/ for Central

Chose this in the latest commit.

<3> Define artifacts used by project
<4> Specific artifact name and version within a specific scope (Guava 31.1-jre for implementation and JUnit 4.13.2 for testImplementation)

As we can see, this theoretical project uses Guava and JUnit as _dependencies_.

Gradle fetches dependencies from the specified _repositories_, in this case, link:https://mvnrepository.com/repos/central[Maven] and link:https://maven.google.com/[Google].
You can <<declaring_repositories.adoc#declaring-repositories,declare repositories>> to tell Gradle where to fetch dependencies locally or remotely.

Repositories offer dependencies in multiple formats.
For information about the formats supported by Gradle, see <<declaring_dependencies.adoc#sec:dependency-types,dependency types>>.

A build engineer can <<declaring_dependencies.adoc#declaring-dependencies,declare dependencies>> for different scopes.

> In the example project, the JUnit was declared only for the `testImplementation` scope which is used only when testing and not at runtime.

Reviewer 2: "the JUnit dependency" or "JUnit", but "the JUnit" feels wrong

Author: Fixed in latest commit.

> As such, one scope is used to compile source code and another to execute tests.

Reviewer 2: I think this sentence requires too much implied knowledge to make sense. The previous one was the same tbh.

Maybe we should replace this with a mention that Gradle uses configurations to scope dependencies and link to the relevant section. Something like [this](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:choosing-configuration).

And then probably use the _configuration_ term in the first sentence of that paragraph, in place of _scopes_.

Author: I made some changes, let me know what you think of the latest commit.
